log-replication/src/main/java/org/corfudb/logreplication/fsm/InSnapshotSyncState.java

```diff
@@ -0,0 +1,200 @@
+package org.corfudb.logreplication.fsm;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import org.corfudb.logreplication.send.SnapshotSender;
+
+import java.util.UUID;
+import java.util.concurrent.Future;
+
+/**
+ * This class represents the InSnapshotSync state of the Log Replication State Machine.
+ *
+ * In this state full logs are being synced to the remote site, based on a snapshot timestamp.
+ */
+@Slf4j
+public class InSnapshotSyncState implements LogReplicationState {
+
+    /*
+     * Log Replication Finite State Machine Instance
+     */
+    private final LogReplicationFSM fsm;
+
+    /*
+     Uniquely identifies the event that caused the transition to this state.
+     This identifier is hold in order to send it back to the application through the DataSender
+     callback, so it can be correlated to the process that triggered the request.
+
+     This is required in the case that a snapshot sync is canceled and another snapshot sync is requested,
+     so the application can discard messages received for the previous snapshot sync, until the new
+     request (event) is handled.
+     */
+    private UUID transitionEventId;
+
+    /*
+     Read and send a snapshot of the data-store.
+     */
+    @Getter
+    @VisibleForTesting
+    private SnapshotSender snapshotSender;
+
+    /*
+     A future on the send, in case we need to cancel the ongoing snapshot sync.
+     */
+    private Future<?> transmitFuture;
+
+    /**
+     * Constructor
+     *
+     * @param logReplicationFSM log replication state machine
+     * @param snapshotSender snapshot sync send (read and send)
+     */
+    public InSnapshotSyncState(LogReplicationFSM logReplicationFSM, SnapshotSender snapshotSender) {
+        this.fsm = logReplicationFSM;
+        this.snapshotSender = snapshotSender;
+    }
+
+    @Override
+    public LogReplicationState processEvent(LogReplicationEvent event) throws IllegalTransitionException {
+        switch (event.getType()) {
+            case SNAPSHOT_SYNC_REQUEST:
+                /*
+                 Cancel ongoing snapshot sync, if it is still in progress.
+                 */
+                cancelSnapshotSync("another snapshot sync request.");
+
+                /*
+                 Set the id of the new snapshot sync request causing the transition.
+
+                 This will be taken onEntry of this state to initiate a snapshot send
+                 for this given request.
+                 */
+                setTransitionEventId(event.getEventID());
+                snapshotSender.reset();
```

Comment:
should it not go back into sync through the event queue. to me this seems like a 
in_sync -> some state like init -> in_sync.  what do you think.